File: FundComponents/ImpactDeathMorbidityComponent.cs
// FUND - Climate Framework for Uncertainty, Negotiation and Distribution
// Copyright (C) 2012 David Anthoff and Richard S.J. Tol
// http://www.fund-model.org
// Licensed under the MIT license
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Esmf;
using Fund.CommonDimensions;

namespace Fund.Components.ImpactDeathMorbidity
{
    public interface IImpactDeathMorbidityState
    {
        IVariable2Dimensional<Timestep, Region, double> dead { get; }
        IVariable2Dimensional<Timestep, Region, double> yll { get; }
        IVariable2Dimensional<Timestep, Region, double> yld { get; }
        IVariable2Dimensional<Timestep, Region, double> deadcost { get; }
        IVariable2Dimensional<Timestep, Region, double> morbcost { get; }
        IVariable2Dimensional<Timestep, Region, double> vsl { get; }
        IVariable2Dimensional<Timestep, Region, double> vmorb { get; }

        IParameter1Dimensional<Region, double> d2ld { get; }
        IParameter1Dimensional<Region, double> d2ls { get; }
        IParameter1Dimensional<Region, double> d2lm { get; }
        IParameter1Dimensional<Region, double> d2lc { get; }
        IParameter1Dimensional<Region, double> d2lr { get; }
        IParameter1Dimensional<Region, double> d2dd { get; }
        IParameter1Dimensional<Region, double> d2ds { get; }
        IParameter1Dimensional<Region, double> d2dm { get; }
        IParameter1Dimensional<Region, double> d2dc { get; }
        IParameter1Dimensional<Region, double> d2dr { get; }

        IParameter2Dimensional<Timestep, Region, double> dengue { get; }
        IParameter2Dimensional<Timestep, Region, double> schisto { get; }
        IParameter2Dimensional<Timestep, Region, double> malaria { get; }
        IParameter2Dimensional<Timestep, Region, double> cardheat { get; }
        IParameter2Dimensional<Timestep, Region, double> cardcold { get; }
        IParameter2Dimensional<Timestep, Region, double> resp { get; }
        IParameter2Dimensional<Timestep, Region, double> diadead { get; }
        IParameter2Dimensional<Timestep, Region, double> hurrdead { get; }
        IParameter2Dimensional<Timestep, Region, double> extratropicalstormsdead { get; }
        IParameter2Dimensional<Timestep, Region, double> population { get; }
        IParameter2Dimensional<Timestep, Region, double> diasick { get; }
        IParameter2Dimensional<Timestep, Region, double> income { get; }

        double vslbm { get; }
        double vslel { get; }
        double vmorbbm { get; }
        double vmorbel { get; }
        double vslypc0 { get; }
        double vmorbypc0 { get; }
    }


    public class ImpactDeathMorbidityComponent
    {
        public void Run(Clock clock, IImpactDeathMorbidityState state, IDimensions dimensions)
        {
            var s = state;
            var t = clock.Current;

            if (clock.IsFirstTimestep)
            {

            }
            else
            {
                foreach (var r in dimensions.GetValues<Region>())
                {
                    double ypc = s.income[t, r] / s.population[t, r] * 1000.0;

                    s.dead[t, r] = s.dengue[t, r] + s.schisto[t, r] + s.malaria[t, r] + s.cardheat[t, r] + s.cardcold[t, r] + s.resp[t, r] + s.diadead[t, r] + s.hurrdead[t, r] + s.extratropicalstormsdead[t, r];
                    if (s.dead[t, r] > s.population[t, r] * 1000000.0)
                        s.dead[t, r] = s.population[t, r] / 1000000.0;

                    s.yll[t, r] = s.d2ld[r] * s.dengue[t, r] + s.d2ls[r] * s.schisto[t, r] + s.d2lm[r] * s.malaria[t, r] + s.d2lc[r] * s.cardheat[t, r] + s.d2lc[r] * s.cardcold[t, r] + s.d2lr[r] * s.resp[t, r];

                    s.yld[t, r] = s.d2dd[r] * s.dengue[t, r] + s.d2ds[r] * s.schisto[t, r] + s.d2dm[r] * s.malaria[t, r] + s.d2dc[r] * s.cardheat[t, r] + s.d2dc[r] * s.cardcold[t, r] + s.d2dr[r] * s.resp[t, r] + s.diasick[t, r];

                    s.vsl[t, r] = s.vslbm * Math.Pow(ypc / s.vslypc0, s.vslel);
                    s.deadcost[t, r] = s.vsl[t, r] * s.dead[t, r] / 1000000000.0;
                    // deadcost:= vyll*ypc*yll/1000000000;

                    s.vmorb[t, r] = s.vmorbbm * Math.Pow(ypc / s.vmorbypc0, s.vmorbel);
                    s.morbcost[t, r] = s.vmorb[t, r] * s.yld[t, r] / 1000000000.0;

                }
            }
        }
    }
}

File: src/components/ImpactDeathMorbidityComponent.jl
@defcomp impactdeathmorbidity begin
    regions = Index()

    dead = Variable(index=[time,regions])
    yll = Variable(index=[time,regions])
    yld = Variable(index=[time,regions])
    deadcost = Variable(index=[time,regions])
    morbcost = Variable(index=[time,regions])
    vsl = Parameter(index=[time,regions])
    vmorb = Parameter(index=[time,regions])

    d2ld = Parameter(index=[regions])
    d2ls = Parameter(index=[regions])
    d2lm = Parameter(index=[regions])
    d2lc = Parameter(index=[regions])
    d2lr = Parameter(index=[regions])
    d2dd = Parameter(index=[regions])
    d2ds = Parameter(index=[regions])
    d2dm = Parameter(index=[regions])
    d2dc = Parameter(index=[regions])
    d2dr = Parameter(index=[regions])

    dengue = Parameter(index=[time,regions])
    schisto = Parameter(index=[time,regions])
    malaria = Parameter(index=[time,regions])
    cardheat = Parameter(index=[time,regions])
    cardcold = Parameter(index=[time,regions])
    resp = Parameter(index=[time,regions])
    diadead = Parameter(index=[time,regions])
    hurrdead = Parameter(index=[time,regions])
    extratropicalstormsdead = Parameter(index=[time,regions])
    population = Parameter(index=[time,regions])
    diasick = Parameter(index=[time,regions])
    # Other sources of death
    dead_other = Parameter(index=[time,regions])
    # Other sources of sickness
    sick_other = Parameter(index=[time,regions])

    function run_timestep(p, v, d, t)

        if !is_first(t)
            for r in d.regions
                v.dead[t, r] = p.dengue[t, r] + p.schisto[t, r] + p.malaria[t, r] + p.cardheat[t, r] + p.cardcold[t, r] + p.resp[t, r] + p.diadead[t, r] + p.hurrdead[t, r] + p.extratropicalstormsdead[t, r] + p.dead_other[t,r]
                if v.dead[t, r] > p.population[t, r] * 1000000.0
                    v.dead[t, r] = p.population[t, r] / 1000000.0
                end

                v.yll[t, r] = p.d2ld[r] * p.dengue[t, r] + p.d2ls[r] * p.schisto[t, r] + p.d2lm[r] * p.malaria[t, r] + p.d2lc[r] * p.cardheat[t, r] + p.d2lc[r] * p.cardcold[t, r] + p.d2lr[r] * p.resp[t, r]

                v.yld[t, r] = p.d2dd[r] * p.dengue[t, r] + p.d2ds[r] * p.schisto[t, r] + p.d2dm[r] * p.malaria[t, r] + p.d2dc[r] * p.cardheat[t, r] + p.d2dc[r] * p.cardcold[t, r] + p.d2dr[r] * p.resp[t, r] + p.diasick[t, r] + p.sick_other[t,r]

                v.deadcost[t, r] = p.vsl[t, r] * v.dead[t, r] / 1000000000.0
                # deadcost:= vyll*ypc*yll/1000000000

                v.morbcost[t, r] = p.vmorb[t, r] * v.yld[t, r] / 1000000000.0
            end
        end
    end
end